helm/v1 from OperatorSDK

anmol372 opened the pull request:
- Added helm/v1 package from OperatorSDK.
Tests run:
	1. Init Project with and without API:
		-> Scaffolding matches OperatorSDK helm operator
	2. create API
		-> created a simple Nginx API, produced helm's boilerplate chart
	3. ran operator with 'helm-operator run' 
- Hardcoded base image version to 'master'.

anmol372 commented:
Scaffolding Created: 
anmolrastogi@anrastog-mac test % tree -d
.
├── config
│   ├── crd
│   │   └── bases
│   ├── default
│   ├── manager
│   ├── prometheus
│   ├── rbac
│   ├── samples
│   └── scorecard
│       ├── bases
│       └── patches
├── helm-charts
   └── nginx
       ├── charts
       └── templates
           └── tests

joelanford reviewed (changes requested):
Looks great! Just a couple organization suggestions and some context about the phase 2 plugin code that I forgot to remind you about.

anmol372 commented:
Modified directory structures and removed sdk phase2 plugin code.
New scaffolding structure: 
anmolrastogi@anrastog-mac ngx % tree -d
.
├── config
│   ├── crd
│   │   └── bases
│   ├── default
│   ├── manager
│   ├── prometheus
│   ├── rbac
│   └── samples
└── helm-charts
    └── nginx
        ├── charts
        └── templates
            └── tests

13 directories

Tests run:
1. Init Project with and without API:
-> Scaffolding matches OperatorSDK helm operator
2. create API
-> created a simple Nginx API, produced helm's boilerplate chart
3. ran operator with 'helm-operator run' and make install+deploy

anmol372 commented on pkg/plugins/v1/scaffolds/internal/templates/dockerfile.go:
```diff
@@ -0,0 +1,49 @@
+}
+
+const dockerfileTemplate = `# Build the manager binary
+FROM quay.io/anrastog/helm-operator:{{.HelmOperatorVersion}}
```
Since I was not aware of any active tags for the image, I added my own image "quay.io/anrastog/helm-operator:canary" to the generated Dockerfile and have been using it for testing.

I'll test the code with quay.io/joelanford/helm-operator:master, and get back to you on what'sbroken.

anmol372 replied:
The `HelmOperatorVersion` is set by `VERSION = $(shell git describe --dirty --tags --always)` in the Makefile right now. We could leave as it is for now.

anmol372 replied:
Update: The `quay.io/joelanford/helm-operator` base image only has `:master` tag, so I have hardcoded HelmOperatorVersion to it as suggested.

anmol372 commented on internal/cmd/run/cmd.go:
```diff
@@ -43,14 +44,16 @@ func NewCmd() *cobra.Command {
 	r.bindFlags(cmd.Flags())
 	cmd.Flags().AddGoFlagSet(zapfs)
+	cmd.Run = func(cmd *cobra.Command, _ []string) {
+		klog.InitFlags(flag.CommandLine)
```
I used this flag initialization, to default to flags set by cli controller, in case 2 controllers for the same operator are launched (one with `make deploy` and the other with cli), but on testing I found that both controllers get stuck on  `Reconcile Errors` `"error":"rendered manifests contain a resource that already exists.`
but one of them(randomly) also carries out reconciler tasks(install/delete) in between printing the error.

So I will revert this change and since this is not a valid behavior for the sdk, we should not need it here.

anmol372 commented on pkg/plugins/v1/scaffolds/internal/templates/makefile.go:
```diff
@@ -0,0 +1,136 @@
+	@{ \
+	set -e ;\
+	mkdir -p bin ;\
+	curl -LO https://github.com/operator-framework/operator-sdk/releases/download/{{ .HelmOperatorVersion }}/helm-operator-{{ .HelmOperatorVersion }}-$(ARCHOPER)-$(OSOPER) ;\
```
Would it be better to have `joelanford/helm-operator` ?
Like: 
`curl -LO https://github.com/joelanford/helm-operator/releases/download/{{ .HelmOperatorVersion }}/helm-operator-{{ .HelmOperatorVersion }}-$(ARCHOPER)-$(OSOPER) ;\`

estroz reviewed (changes requested):
Some early suggestions:
- `pkg/plugins/internal/kubebuilder -> pkg/plugins/internal`
- Since there may be more plugins than just the base (phase 1) helm plugin, ex. phase 2 plugins: `pkg/plugins/v1 -> pkg/plugins/helm/v1`

joelanford commented:
@estroz I made conflicting suggestions to @anmol372 in previous comments, so let's discuss.

> pkg/plugins/internal/kubebuilder -> pkg/plugins/internal

I think we should keep the kubebuilder-specific libraries isolated and together in their own subdirectory just to make it more obvious what's from kubebuilder and what's original to this repo. 

> Since there may be more plugins than just the base (phase 1) helm plugin, ex. phase 2 plugins: pkg/plugins/v1 -> pkg/plugins/helm/v1

I suggested the exact opposite :rofl: 

My logic was, "This is the helm-operator repo, therefore all plugins will be helm plugins, so no need for a helm subdirectory". Re: possibility of phase 2 plugins, I think I envision those being maintained in a separate repo (e.g. `operator-sdk` or a 3rd repo).

I suppose the _one_ case I can think of is a helm-specific phase 2 plugin that would be opt-in, but I can't think of any examples of what that would be off the top of my head.

estroz commented:
>I think we should keep the kubebuilder-specific libraries isolated and together in their own subdirectory just to make it more obvious what's from kubebuilder and what's original to this repo.

That's reasonable.

>I suppose the one case I can think of is a helm-specific phase 2 plugin that would be opt-in, but I can't think of any examples of what that would be off the top of my head.

This is the only scenario I am considering. I can't either currently, just planning for that future scenario now.

anmol372 commented:
@estroz , @joelanford 
To sum up, I leave `pkg/plugins/internal/kubebuilder` and `pkg/plugins/v1` as they are now, right?

estroz commented:
@anmol372 yeah, we can come back to pathing later/never.

anmol372 commented on Makefile:
```diff
@@ -12,7 +12,8 @@ endif
 REPO = $(shell go list -m)
-VERSION = $(shell git describe --dirty --tags --always)
+#TODO (anrastog): set version to repo build/tag after v1 plugin is available in master.
+VERSION = master
```
Hardcoded the image version to master as that is the only tag that exists for quay.io/joelanford/helm-operator right now.

estroz commented on Makefile:
```diff
@@ -45,14 +46,15 @@ vet:
 # Build the docker image
 docker-build:
-	docker build . -t ${IMG}
+	docker build . -t ${IMG}:$(VERSION)
```
Wouldn't it be better to call `make docker-build IMG=quay.io/joelanford/helm-operator:$VERSION`? Appending `$(VERSION)` is unexpected behavior IMO.

estroz commented on main.go:
```diff
@@ -19,21 +19,34 @@ package main
+	c, err := cli.New(
+		cli.WithCommandName("helm-operator"),
+		cli.WithPlugins(
+			&pluginv1.Plugin{},
```
So `helm-operator` is going to have scaffolding subcommands? I'd prefer to split these CLI changes into a separate PR, and only add the plugin code in `pkg/plugins`.

anmol372 replied:
I can start a separate PR for the CLI changes.

What I understand from a previous discussion with Joe, is that this repo will be scaffolding the required files to build a `helm-operator` and once it's ready `operator-sdk` will import and run the vanilla helm-operator plugin from this repo, and separately apply its own phase 2 plugins (additional sub-commands), which are effectively overlaid on the base.

estroz reviewed (changes requested):
Assuming most of this was copied directly from `operator-sdk/internal/plugins/helm`, so everything in `pkg/plugins` LGTM. Otherwise, one a concern about adding scaffolding CLI changes in this PR.

estroz commented on Makefile:
```diff
@@ -45,14 +46,15 @@ vet:
 # Build the docker image
 docker-build:
-	docker build . -t ${IMG}
+	docker build . -t quay.io/joelanford/helm-operator:$(VERSION)
```
I meant for this not to be changed at all, but I suppose this is fine for now because the only tag that can be built for export is this one.

estroz reviewed (approved):
/lgtm
